DirectProgramming/DPC++FPGA/include/MemoryTransfers.hpp

```diff
@@ -0,0 +1,296 @@
+          int rows,            // Number of rows of the matrix
+          int columns,         // Number of columns of the matrix
+          int numElemPerBank,  // Number of TT elements per DDR burst access
+          int matrixCount,     // Number of matrices to read from the buffer
+          typename matrixPipe  // Output matrix pipe
+          >
+void MatrixReadFromDDRToPipe(
+    TT* MatrixPtr,  // Input matrix buffer
+    typename std::enable_if<(rows % numElemPerBank) == 0>::type* = 0) {
+  // Number of DDR burst reads of numElemPerBank required to read a full column
+  constexpr int kLoopIterPerColumn = rows / numElemPerBank;
+  // Number of DDR burst reads of numElemPerBank to read all the matrices
+  constexpr int kLoopIter = kLoopIterPerColumn * columns * matrixCount;
+  // Size in bits of the loop iterator over kLoopIter iterations
+  constexpr int kLoopIterBitSize = BitsForMaxValue<kLoopIter + 1>();
+
+  sycl::device_ptr<TT> MatrixPtrDevice(MatrixPtr);
+
+  [[intel::initiation_interval(1)]]  // NO-FORMAT: Attribute
+  for (ac_int<kLoopIterBitSize, false> li = 0; li < kLoopIter; li++) {
+    pipeTable<numElemPerBank, TT> DDRRead;
+    // Perform the DDR burst read of numElemPerBank elements
+    UnrolledLoop<numElemPerBank>([&](auto k) {
+      DDRRead.elem[k] = MatrixPtrDevice[(int)(li)*numElemPerBank + k];
+    });
+
+    matrixPipe::write(DDRRead);
+  }  // end of li
+}
+
+/*
+  Read "matrixCount" matrices of type TT from DDR by bursts of numElemPerBank
+  elements, and write the matrices to the "matrixPipe" pipe numElemPerBank by
+  numElemPerBank.
+  This implementation is used for matrices that have a number of rows that is
+  not a multiple of the number of elements per DDR burst read (numElemPerBank).
+  Another version of this function is written above and will be selected at
+  compile time if the row count is a multiple numElemPerBank.
+*/
+template <typename TT,         // Datatype of the elements of the matrix
+          int rows,            // Number of rows of the matrix
+          int columns,         // Number of columns of the matrix
+          int numElemPerBank,  // Number of TT elements per DDR burst access
+          int matrixCount,     // Number of matrices to read from the buffer
+          typename matrixPipe  // Output matrix pipe
+          >
+void MatrixReadFromDDRToPipe(
+    TT* MatrixPtr,  // Input matrix buffer
+    typename std::enable_if<(rows % numElemPerBank) != 0>::type* = 0) {
+  // Number of DDR burst reads of numElemPerBank required to read a full column
+  constexpr int kLoopIterPerColumn = rows / numElemPerBank + 1;
+  // Number of DDR burst reads of numElemPerBank to read all the matrices
+  constexpr int kLoopIter = kLoopIterPerColumn * columns * matrixCount;
+  // Size in bits of the loop iterator over kLoopIter iterations
+  constexpr int kLoopIterBitSize = BitsForMaxValue<kLoopIter + 1>();
+
+  sycl::device_ptr<TT> MatrixPtrDevice(MatrixPtr);
```

Comment:
code formatting for local variables.
--- Last time I'll say it, you get the point.